// kenat/src/nigs.js
export const orthodoxMonthlydays = {
  ledeta_mariam: {
    key: "ledeta_mariam",
    name: {
      english: "Ldata Mariam, The nativity of Our Holy Mother Virgin Mariam",
      amharic: "ልደታ ማርያም",
    },
    description: {
      english: "Commemoration of the birth of St. Mary.",
      amharic: "የቅድስት ድንግል ማርያም ልደት መታሰቢያ።",
    },
    recuringDate: 1,
    major: true,
    negs: 9,
  },
  archangel_raguel: {
    key: "archangel_raguel",
    name: {
      english: "Archangel Raguel",
      amharic: "ራጉኤል",
    },
    description: {
      english: "Commemoration of Archangel Raguel.",
      amharic: "የመላኩ ራጉኤል መታሰቢያ።",
    },
    recuringDate: 1,
    major: true,
    negs: null,
  },
  bartholomew_the_apostle: {
    key: "bartholomew_the_apostle",
    name: {
      english: "Bartholomew the Apostle",
      amharic: "ሐዋርያው በርተሎሜዎስ",
    },
    description: {
      english: "Commemoration of Bartholomew the Apostle.",
      amharic: "የሐዋርያው በርተሎሜዎስ መታሰቢያ።",
    },
    recuringDate: 1,
    major: true,
    negs: null,
  },
  prophet_elijah_ginbot: {
    key: "prophet_elijah_ginbot",
    name: {
      english: "Prophet Elijah",
      amharic: "ነብዩ ኤልያስ",
    },
    description: {
      english: "Commemoration of Prophet Elijah.",
      amharic: "የነብዩ ኤልያስ መታሰቢያ።",
    },
    recuringDate: 1,
    major: true,
    negs: 4,
  },
  kidus_yohannes: {
    key: "kidus_yohannes",
    name: {
      english: "Kidus Yohannes",
      amharic: "ቅዱስ ዮሐንስ",
    },
    description: {
      english: "Commemoration of Kidus Yohannes.",
      amharic: "የቅዱስ ዮሐንስ መታሰቢያ።",
    },
    recuringDate: 1,
    major: false,
    negs: 1,
  },


  abba_guba: {
    key: "abba_guba",
    name: {
      english: "Abba Guba, one of the Nine Saints",
      amharic: "አባ ጉባ",
    },
    description: {
      english: "Commemoration of Abba Guba, one of the Nine Saints.",
      amharic: "ከዘጠኙ ቅዱሳን አንዱ የአባ ጉባ መታሰቢያ።",
    },
    recuringDate: 2,
    major: true,
    negs: null,
  },

  thaddeus_the_apostle: {
    name: {
      english: "Thaddeus the Apostle",
      amharic: "ሐዋርያው ታድዮስ",
    },
    recuringDate: 2,
    defaultDescription: {
        english: "Commemoration of Thaddeus the Apostle.",
        amharic: "ሐዋርያው ታድዮስ መታሰቢያ",
    },
    events: [
      {
        key: "thaddeus_the_apostle_major",
        description: { english: "Commemoration of Thaddeus the Apostle.", amharic: "የሐዋርያው ታድዮስ መታሰቢያ።" },
        major: true,
        negs: 11,
      },
      {
        key: "thaddeus_apostle_minor",
        description: { english: "Commemoration of Thaddeus the Apostle.", amharic: "የሐዋርያው ታድዮስ መታሰቢያ።" },
        major: false,
        negs: 13,
      }
    ]
  },

  st_john_the_baptist_martyrdom: {
    key: "st_john_the_baptist_martyrdom",
    name: {
      english: "The Martyrdom (Beheading) of St. John the Baptist",
      amharic: "የቅዱስ ዮሐንስ መጥምቅ አንገት መቁረጥ",
    },
    description: {
      english:
        "Commemoration of the Martyrdom (Beheading) of St. John the Baptist.",
      amharic: "የቅዱስ ዮሐንስ መጥምቅ አንገት መቁረጥ መታሰቢያ።",
    },
    recuringDate: 2,
    major: false,
    negs: 1,
  },

  beata: {
    key: "beata",
    name: {
      english:
        "Beata: The Entrance of the three years old Virgin Mariam into the Temple",
      amharic: "በአታለማርያም",
    },
    description: {
      english: "Commemoration of St. Mary's entrance into the Temple.",
      amharic: "የቅድስት ድንግል ማርያም ወደ ቤተመቅደስ መግባት መታሰቢያ።",
    },
    recuringDate: 3,
    major: true,
    negs: 4,
  },
  abune_zena_markos: {
    key: "abune_zena_markos",
    name: {
      english: "Abune Zena Markos",
      amharic: "አቡነ ዜና ማርቆስ",
    },
    description: {
      english: "Commemoration of Abune Zena Markos.",
      amharic: "የአቡነ ዜና ማርቆስ መታሰቢያ።",
    },
    recuringDate: 3,
    major: true,
    negs: null,
  },
  archangel_phanuel: {
    key: "archangel_phanuel",
    name: {
      english: "Archangel Phanuel",
      amharic: "ቅዱስ ፋኑኤል",
    },
    description: {
      english: "Commemoration of Archangel Phanuel.",
      amharic: "የመላኩ ፋኑኤል መታሰቢያ።",
    },
    recuringDate: 3,
    major: true,
    negs: 13,
  },
  saint_neakuto_leab: {
    key: "saint_neakuto_leab",
    name: {
      english: "Saint Neakuto Leab",
      amharic: "ቅዱስ ነአኩቶ ለአብ",
    },
    description: {
      english: "Commemoration of Saint Neakuto Leab.",
      amharic: "የቅዱስ ነአኩቶ ለአብ መታሰቢያ።",
    },
    recuringDate: 3,
    major: true,
    negs: null,
  },

  johannis_welde_negedguad: {
    key: "johannis_welde_negedguad",
    name: {
      english: "John the son of the Thunder (Yohannis Welde Negedguad)",
      amharic: "ዮሐንስ ወልደ ነጎድጓድ",
    },
    description: {
      english: "Commemoration of John the son of the Thunder.",
      amharic: "የዮሐንስ ወልደ ነጎድጓድ መታሰቢያ።",
    },
    recuringDate: 4,
    major: true,
    negs: 5,
  },

  andrew_the_apostle: {
    key: "andrew_the_apostle",
    name: {
      english: "Andrew the Apostle",
      amharic: "ሐዋርያው እንድርያስ",
    },
    description: {
      english: "Commemoration of Andrew the Apostle.",
      amharic: "የሐዋርያው እንድርያስ መታሰቢያ።",
    },
    recuringDate: 4,
    major: true,
    negs: 4,
  },

  petros_we_paulos: {
    key: "petros_we_paulos",
    name: {
      english: "Petros we Paulos (Peter and Paul)",
      amharic: "ጴጥሮስወ ጳውሎስ",
    },
    description: {
      english: "Commemoration of Saints Peter and Paul.",
      amharic: "የቅዱሳን ጴጥሮስና ጳውሎስ መታሰቢያ።",
    },
    recuringDate: 5,
    major: true,
    negs: 11,
  },
  abuna_gebre_menfes_kidus: {
    key: "abuna_gebre_menfes_kidus",
    name: {
      english: "Abuna Gebre Menfes Kidus",
      amharic: "ገብረ መንፈስ ቅዱስ",
    },
    description: {
      english: "Commemoration of Abuna Gebre Menfes Kidus.",
      amharic: "የአቡነ ገብረ መንፈስ ቅዱስ መታሰቢያ።",
    },
    recuringDate: 5,
    major: true,
    negs: 2,
  },
  abune_arone: {
    key: "abune_arone",
    name: {
      english: "Abune Arone",
      amharic: "አቡነ አሮን",
    },
    description: {
      english: "Commemoration of Abune Arone.",
      amharic: "የአቡነ አሮን መታሰቢያ።",
    },
    recuringDate: 5,
    major: true,
    negs: null,
  },

  iyyasu: {
    key: "iyyasu",
    name: {
      english: "Iyyasus",
      amharic: "ኢየሱስ",
    },
    description: {
      english: "Commemoration of Iyyasus.",
      amharic: "የኢየሱስ መታሰቢያ።",
    },
    recuringDate: 6,
    major: true,
    negs: 5,
  },
  dabra_quesqam_mariam: {
    key: "dabra_quesqam_mariam",
    name: {
      english: "Dabra Quesqam Mariam",
      amharic: "ቁስቋም ማርያም",
    },
    description: {
      english: "Commemoration of Dabra Quesqam Mariam.",
      amharic: "የቁስቋም ማርያም መታሰቢያ።",
    },
    recuringDate: 6,
    major: true,
    negs: 3,
  },
  saint_arsema: {
    key: "saint_arsema",
    name: {
      english: "Saint Arsema",
      amharic: "አርሴማ",
    },
    description: {
      english: "Commemoration of Saint Arsema.",
      amharic: "የቅድስት አርሴማ መታሰቢያ።",
    },
    recuringDate: 6,
    major: true,
    negs: 4,
  },

  holy_trinity: {
    key: "holy_trinity",
    name: {
      english: "Holy Trinity",
      amharic: "አጋዕዝተ አለም ስላሴ",
    },
    description: {
      english: "Commemoration of the Holy Trinity.",
      amharic: "የአጋዕዝተ ዓለም ስላሴ መታሰቢያ።",
    },
    recuringDate: 7,
    major: true,
    negs: 5,
  },

  cherubim: {
    key: "cherubim",
    name: {
      english: "Cherubim (Four Living Creatures)",
      amharic: "ኪሩቤል አርባእቱ እንስሳ",
    },
    description: {
      english: "Commemoration of the Cherubim, the four living creatures.",
      amharic: "የኪሩቤል አርባእቱ እንስሳ መታሰቢያ።",
    },
    recuringDate: 8,
    major: true,
    negs: 3,
  },
  saint_matthias_apostle: {
    key: "saint_matthias_apostle",
    name: {
      english: "Saint Matthias the Apostle",
      amharic: "ሐዋርያው ማትያስ",
    },
    description: {
      english: "Commemoration of Saint Matthias the Apostle.",
      amharic: "የሐዋርያው ማትያስ መታሰቢያ።",
    },
    recuringDate: 8,
    major: true,
    negs: 7,
  },
  abune_kiros: {
    key: "abune_kiros",
    name: {
      english: "Abune Kiros",
      amharic: "አቡነ ኪሮስ",
    },
    description: {
      english: "Commemoration of Abune Kiros.",
      amharic: "የአቡነ ኪሮስ መታሰቢያ።",
    },
    recuringDate: 8,
    major: true,
    negs: 3,
  },
  abba_banuda: {
    key: "abba_banuda",
    name: {
      english: "Abba Banuda",
      amharic: "አባ ባኑዳ",
    },
    description: {
      english: "Commemoration of Abba Banuda.",
      amharic: "የአባ ባኑዳ መታሰቢያ።",
    },
    recuringDate: 8,
    major: false,
    negs: null,
  },

  saint_thomas_apostle: {
    key: "saint_thomas_apostle",
    name: {
      english: "Saint Thomas the Apostle",
      amharic: "ሐዋርያው ቶማስ",
    },
    description: {
      english: "Commemoration of Saint Thomas the Apostle.",
      amharic: "የሐዋርያው ቶማስ መታሰቢያ።",
    },
    recuringDate: 9,
    major: true,
    negs: 2,
  },
  abune_isitinifase_kirisitosi: {
    key: "abune_isitinifase_kirisitosi",
    name: {
      english: "Abune Isitinifase Kirisitosi",
      amharic: "አቡነ እስትንፋሰ ክርስቶስ",
    },
    description: {
      english: "Commemoration of Abune Isitinifase Kirisitosi.",
      amharic: "የአቡነ እስትንፋሰ ክርስቶስ መታሰቢያ።",
    },
    recuringDate: 9,
    major: true,
    negs: null,
  },
  three_eighteen_holy_fathers_nice: {
    key: "three_eighteen_holy_fathers_nice",
    name: {
      english: "The 318 holy fathers assembled in the city of Nice",
      amharic: "ሰልስቱ ምዕት",
    },
    description: {
      english:
        "Commemoration of the 318 holy fathers assembled in the city of Nice.",
      amharic: "በኒቂያ ከተማ የተሰበሰቡት 318 ቅዱሳን አባቶች መታሰቢያ።",
    },
    recuringDate: 9,
    major: true,
    negs: null,
  },

  holy_cross: {
    key: "holy_cross",
    name: {
      english: "Holy Cross (Masqal)",
      amharic: "መስቀለ ክርስቶስ",
    },
    description: {
      english:
        "Commemoration of the Holy Cross (Masqal)/the finding of true cross.",
      amharic: "የመስቀለ ክርስቶስ መታሰቢያ።",
    },
    recuringDate: 10,
    major: true,
    negs: 7,
  },
  tsedeniya_mariyami: {
    key: "tsedeniya_mariyami",
    name: {
      english: "Ts'edeniya Mariyami",
      amharic: "ፀደንያ ማርያም",
    },
    description: {
      english: "Commemoration of Ts'edeniya Mariyami.",
      amharic: "የፀደንያ ማርያም መታሰቢያ።",
    },
    recuringDate: 10,
    major: true,
    negs: 1,
  },
  simon_the_zealot: {
    key: "simon_the_zealot",
    name: {
      english: "Simon the Zealot",
      amharic: "ስምኦን ቀኖናዊ ሐዋርያ",
    },
    description: {
      english: "Commemoration of Simon the Zealot.",
      amharic: "የስምኦን ቀኖናዊ ሐዋርያ መታሰቢያ።",
    },
    recuringDate: 10,
    major: true,
    negs: 11,
  },

  saint_hanna: {
    key: "saint_hanna",
    name: {
      english: "Saint Hanna, the mother of our holy Mother",
      amharic: "ሐና",
    },
    description: {
      english: "Commemoration of Saint Hanna, the mother of Our Holy Mother.",
      amharic: "የቅድስት ሐና (የእመቤታችን እናት) መታሰቢያ።",
    },
    recuringDate: 11,
    major: true,
    negs: 11,
  },
  saint_joachim: {
    key: "saint_joachim",
    name: {
      english: "Saint Joachim, the father of our holy Mother",
      amharic: "ኢያቄም",
    },
    description: {
      english: "Commemoration of Saint Joachim, the father of Our Holy Mother.",
      amharic: "የቅዱስ ኢያቄም (የእመቤታችን አባት) መታሰቢያ።",
    },
    recuringDate: 11,
    major: false,
    negs: null,
  },
  saint_yared: {
    key: "saint_yared",
    name: {
      english: "Saint Yared",
      amharic: "ቅዱስ ያሬድ",
    },
    description: {
      english: "Commemoration of Saint Yared.",
      amharic: "የቅዱስ ያሬድ መታሰቢያ።",
    },
    recuringDate: 11,
    major: true,
    negs: 9,
  },
  abune_hara: {
    key: "abune_hara",
    name: {
      english: "Abune Hara",
      amharic: "አቡነ ሐራ",
    },
    description: {
      english: "Commemoration of Abune Hara.",
      amharic: "የአቡነ ሐራ መታሰቢያ።",
    },
    recuringDate: 11,
    major: false,
    negs: null,
  },

  archangel_michael: {
    key: "archangel_michael",
    name: {
      english: "Archangel Michael",
      amharic: "ቅዱስ ሚካኤል",
    },
    description: {
      english: "Commemoration of Archangel Michael.",
      amharic: "የቅዱስ ሚካኤል መታሰቢያ።",
    },
    recuringDate: 12,
    major: true,
    negs: [3, 12],
  },
  matthew_apostle: {
    key: "matthew_apostle",
    name: {
      english: "Matthew the Apostle",
      amharic: "ሐዋርያ ማቴዎስ",
    },
    description: {
      english: "Commemoration of Matthew the Apostle.",
      amharic: "የሐዋርያ ማቴዎስ መታሰቢያ።",
    },
    recuringDate: 12,
    major: true,
    negs: null,
  },
  abba_samuel_waldeba: {
    key: "abba_samuel_waldeba",
    name: {
      english: "Abba Samuel from Waldeba",
      amharic: "አባ ሳሙኤል",
    },
    description: {
      english: "Commemoration of Abba Samuel from Waldeba.",
      amharic: "ከዋልድባ የአባ ሳሙኤል መታሰቢያ።",
    },
    recuringDate: 12,
    major: true,
    negs: 4,
  },

  god_the_father: {
    key: "god_the_father",
    name: {
      english: "Egziabher Abe (God the Father)",
      amharic: "እግዚአብሔር አብ",
    },
    description: {
      english: "Commemoration of God the Father.",
      amharic: "የእግዚአብሔር አብ መታሰቢያ።",
    },
    recuringDate: 13,
    major: true,
    negs: 3,
  },
  archangel_raphael: {
    key: "archangel_raphael",
    name: {
      english: "Archangel Raphael",
      amharic: "ቅዱስ ሩፋኤል",
    },
    description: {
      english: "Commemoration of Archangel Raphael.",
      amharic: "የቅዱስ ሩፋኤል መታሰቢያ።",
    },
    recuringDate: 13,
    major: true,
    negs: 4,
  },
  saint_abba_zara_buruk: {
    key: "saint_abba_zara_buruk",
    name: {
      english: "Saint Abba Zar'a-Buruk",
      amharic: "አቡነ ዘርአ ብሩክ",
    },
    description: {
      english:
        "Commemoration of Saint Abba Zar'a-Buruk-departed from this world.",
      amharic: "የቅዱስ አባ ዘርአ ብሩክ እረፍታቸው መታሰቢያ።",
    },
    recuringDate: 13,
    major: true,
    negs: 5,
  },

  // 14. Abuna Aragwi
  abuna_aragwi: {
    key: "abuna_aragwi",
    name: {
      english: "Abuna Aragwi-one of the Nine Saints",
      amharic: "አቡነ አረጋዊ ",
    },
    description: {
      english: "Commemoration of Abuna Aragwi, one of the Nine Saints",
      amharic: "ከዘጠኙ ቅዱሳን አንዱ የአቡነ አረጋዊ መታሰቢያ።",
    },
    recuringDate: 14,
    major: true,
    negs: 2,
  },
  gabra_krestos_the_hermit: {
    key: "gabra_krestos_the_hermit",
    name: {
      english: "Gabra Krestos the hermit",
      amharic: "ገብረ ክርስቶስ/ገብረ መርአዊ",
    },
    description: {
      english: "Commemoration of  Gabra Krestos the hermit.",
      amharic: "የገብረ ክርስቶስ ገዳማዊ መታሰቢያ።",
    },
    recuringDate: 14,
    major: true,
    negs: 9,
  },

  cyriacus_julietta: {
    key: "cyriacus_julietta",
    name: {
      english: "St. Cyriacus and St. Julietta",
      amharic: "ቂርቆስና እየሉጣ",
    },
    description: {
      english: "Commemoration of St. Cyriacus and St. Julietta.",
      amharic: "የቅዱስ ቂርቆስና የቅድስት እየሉጣ መታሰቢያ።",
    },
    recuringDate: 15,
    major: true,
    negs: [11, 5],
  },
  saint_stefanos_migration_of_his_body: {
    key: "saint_stefanos_migration_of_his_body",
    name: {
      english: "Saint Stefanos",
      amharic: "ቅዱስ እስጢፋኖስ ",
    },
    description: {
      english:
        "Commemoration  of Saint Estifanos the First Martyr migration of his body",
      amharic: "የቅዱስ እስጢፋኖስ ሰማዕት (ቀዳሚ ሰማዕትና ሊቀ ዲያቆናት) አካል ሽግግራቸው መታሰቢያ",
    },
    recuringDate: 15,
    major: false,
    negs: null,
  },

  kidane_meheret: {
    key: "kidane_meheret",
    name: {
      english: "Kidane Meheret - Our Lady, Covenant of Mercy",
      amharic: "ኪዳነ ምህረት",
    },
    description: {
      english: "Commemoration of Kidane Meheret (Covenant of Mercy).",
      amharic: "የኪዳነ ምህረት መታሰቢያ።",
    },
    recuringDate: 16,
    major: true,
    negs: [12, 6],
  },
  johannis_welde_negedguad_erget: {
    key: "johannis_welde_negedguad_16",
    name: {
      english: "John the son of the Thunder (Yohannis Welde Negedguad)",
      amharic: "ዮሐንስ ወልደ ነጎድጓድ",
    },
    description: {
      english:
        "Commemoration of John, the Son of Thunder his taken away to heaven.",
      amharic: "የዮሐንስ ወልደ ነጎድጓድ ወደ ሰማይ መውሰዳቸው መታሰቢያ።",
    },
    recuringDate: 16,
    major: false,
    negs: 11,
  },

  saint_stefanos: {
    key: "saint_stefanos",
    name: {
      english:
        "Saint Stefanos (Stephen the Martyr) - the First Martyr and Archdeacon)",
      amharic: "ቅዱስ እስጢፋኖስ ፣ ሐዋርያውያዕቆብ ወልደ ዘብዲዮስ ፣ አባ ገሪማ",
    },
    description: {
      english:
        "Commemoration of Saint Stephen the Martyr (First Martyr and Archdeacon) ordination date",
      amharic: "የቅዱስ እስጢፋኖስ ሰማዕት (ቀዳሚ ሰማዕትና ሊቀ ዲያቆናት)  የክህነት ዘመቻቸው መታሰቢያ",
    },
    recuringDate: 17,
    major: true,
    negs: 2,
  },
  james_the_apostle_17: {
    key: "james_the_apostle_17",
    name: {
      english: " James the apostle",
      amharic: " ሐዋርያው ያዕቆብ ወልደ ዘብዲዮስ",
    },
    description: {
      english: "Commemoration of James the Apostle.",
      amharic: "የሐዋርያው ያዕቆብ ወልደ ዘብዲዮስ መታሰቢያ።",
    },
    recuringDate: 17,
    major: true,
    negs: 8,
  },
  abba_gerima: {
    key: "abba_gerima",
    name: {
      english: "Abba Gerima (one of the Nine saints)",
      amharic: "አባ ገሪማ",
    },
    description: {
      english: "Commemoration of  Abba Gerima (one of the Nine Saints).",
      amharic: "የአባ ገሪማ (ከዘጠኙ ቅዱሳን አንዱ) መታሰቢያ።",
    },
    recuringDate: 17,
    major: true,
    negs: null,
  },

  philip_the_apostle: {
    key: "philip_the_apostle",
    name: {
      english: "Philip the Apostle /Abune Ewostatewos",
      amharic: "ሐዋርያው ፊሊጶስ ",
    },
    description: {
      english: "Commemoration of Philip the Apostle.",
      amharic: "የሐዋርያው ፊሊጶስ መታሰቢያ።",
    },
    recuringDate: 18,
    major: true,
    negs: 3,
  },
  abune_ewostatewos: {
    key: "abune_ewostatewos",
    name: {
      english: "Abune_Ewostatewos",
      amharic: "አቡነ ኢዩስጣቲዮስ",
    },
    description: {
      english: "Commemoration of  Abune Ewostatewos.",
      amharic: "የአቡነ ኢዩስጣቲዮስ መታሰቢያ።",
    },
    recuringDate: 18,
    major: true,
    negs: 1,
  },
  saint_georgis_scattering_of_his_bone: {
    key: "saint_georgis_scattering_of_his_bone",
    name: {
      english: "Saint Georgis",
      amharic: "ቅዱስ ጊዮርጊስ",
    },
    description: {
      english:
        " Feast of the Great Martyr Saint George, the scattering of his bone",
      amharic: "ቅዱስ ጊዮርጊስ የአጥንቱ ፍልሰት መታሰቢያ",
    },
    recuringDate: 24,
    major: false,
    negs: 5,
  },
  james_the_apostle_18: {
    key: "james_the_apostle_18",
    name: {
      english: " James the apostle",
      amharic: " ሐዋርያው ያዕቆብ ወልደ ዘብዲዮስ",
    },
    description: {
      english: "Commemoration of James the Apostle.",
      amharic: "የሐዋርያው ያዕቆብ ወልደ ዘብዲዮስ መታሰቢያ።",
    },
    recuringDate: 18,
    major: false,
    negs: 11,
  },

  // 19. Gabriel the Archangel
  gabriel_archangel_rescue_of_the_three_youths: {
    key: "gabriel_archangel_rescue_of_the_three_youths",
    name: {
      english: "Gabriel the Archangel",
      amharic: "ቅዱስ ገብርኤል",
    },
    description: {
      english:
        "Commemoration of Archangel Gabriel is dedicated to the rescue of the three youths (Sidraq, Misaq, and Abdenago).",
      amharic: "ሊቀ መላእክት ገብርኤል ሦስቱን ወጣቶች (ሲድራቅ፣ ሚሳቅና አብደናጎን) ማዳን መታሰቢያ",
    },
    recuringDate: 19,
    major: true,
    negs: 4,
  },
  gabriel_archangel_rescue_of_rescue_of_quirqos_and_his_mother: {
    key: "gabriel_archangel_rescue_of_rescue_of_quirqos_and_his_mother",
    name: {
      english: "Gabriel the Archangel",
      amharic: "ቅዱስ ገብርኤል",
    },
    description: {
      english:
        "Commemoration of Archangel Gabriel is dedicated to the rescue of rescue of Quirqos and his mother, Saint Iyeluta (Julitta). .",
      amharic: "የእግዚአብሔር መልአክ ቅዱስ ገብርኤል ቂርቆስን እና እናቱን ቅድስት ኢየሉጣን ማዳን መታሰቢያ",
    },
    recuringDate: 19,
    major: false,
    negs: 11,
  },

  hnstata_betkerestyan: {
    key: "hnstata_betkerestyan",
    name: {
      english: "Hnstata Betkerestyan",
      amharic: "ሕንፀተ ቤተ ክርስቲያን",
    },
    description: {
      english:
        "Commemoration of the building of the church in the name of Our Holy Virgin Mary.",
      amharic:
        "ሄነሳተ ቤተ ክርስቲያን - ለቅድስት ድንግል ማርያም የተሰጠች የመጀመሪያዋ ቤተ ክርስቲያን በፊልጵስዩስ መመሥረት",
    },
    recuringDate: 20,
    major: true,
    negs: 10,
  },

  egzeet_na_maryam_departure: {
    key: "egzeet_na_maryam_departure",
    name: {
      english: "Egze'et-na Maryam",
      amharic: "ቅድስት ድንግል ማርያም",
    },
    description: {
      english: "The departure of Our Holy Mother.",
      amharic: "የቅድስት ድንግል ማርያም ዕርገት",
    },
    recuringDate: 21,
    major: false,
    negs: 5,
  },
  egzeet_na_maryam_debre_mitmak: {
    key: "egzeet_na_maryam_debre_mitmak",
    name: {
      english: "Egze'et-na Maryam",
      amharic: "ቅድስት ድንግል ማርያም",
    },
    description: {
      english:
        "The appearance of Our Holy Mother Virgin Mariam in the church of Debre Mitmak",
      amharic: "የቅድስት ድንግል ማርያም በደብረ ምጥማቅ ቤተ ክርስቲያን መገለጥ።",
    },
    recuringDate: 21,
    major: false,
    negs: 9,
  },
  egzeet_na_maryam_prayed_at_golgota: {
    key: "egzeet_na_maryam_prayed_at_golgota",
    name: {
      english: "Egze'et-na Maryam",
      amharic: "ቅድስት ድንግል ማርያም",
    },
    description: {
      english: "Sene Golgota - Our Holy Mother Virgin Mariam prayed at Golgota",
      amharic: "ሰኔ ጎልጎታ - ቅድስት ድንግል ማርያም በጎልጎታ ጸለየች",
    },
    recuringDate: 21,
    major: false,
    negs: 10,
  },

  // 22. Archangel Uriel

  archangel_uriel: {
    key: "archangel_uriel",
    name: {
      english: "Archangel Uriel",
      amharic: "ቅዱስ ኡራኤል",
    },
    description: {
      english: "Archangel Uriel received covenant and ordained as an Archangel",
      amharic: "ሊቀ መላእክት ዑራኤል ቃል ኪዳን ተቀብሎ ሊቀ መልአክ ሆኖ ተሾመ።",
    },
    recuringDate: 22,
    major: false,
    negs: 11,
  },
  saint_lukas: {
    key: "saint_lukas",
    name: {
      english: "Saint Lukas",
      amharic: "ቅዱስ ሉቃስ",
    },
    description: {
      english: "Saint Mark departed from this world",
      amharic: "ቅዱስ ማርቆስ ከዚህ ዓለም በሞት ተለየ",
    },
    recuringDate: 22,
    major: true,
    negs: 2,
  },
  saint_daqsyos: {
    key: "saint_daqsyos",
    name: {
      english: "Saint Daqsyos",
      amharic: "ቅዱስ ደቅስዮስ",
    },
    description: {
      english: "Saint_Daqsyos",
      amharic: "ቅዱስ ደቅስዮስ",
    },
    recuringDate: 22,
    major: true,
    negs: 4,
  },
  bisrate_gebriel: {
    key: "bisrate_gebriel",
    name: {
      english: "bisrate gebriel ",
      amharic: "ብስራተ ገብርኤል",
    },
    description: {
      english: "Commemoration of the Annunciation",
      amharic: "ብስራተ ገብርኤል",
    },
    recuringDate: 22,
    major: true,
    negs: 4,
  },

  saint_georgis_of_Lydda: {
    key: "saint_georgis_of_Lydda",
    name: {
      english: "Saint Georgis - George of Lydda",
      amharic: "ቅዱስ ጊዮርጊስ",
    },
    description: {
      english: " Saint Georgis - George of Lydda",
      amharic: "ቅዱስ ጊዮርጊስ ",
    },
    recuringDate: 23,
    major: false,
    negs: 8,
  },
  saint_georgis_the_battle_of_adwa: {
    key: "saint_georgis_the_battle_of_adwa",
    name: {
      english: "Saint Giorgis (George)-The Battle of Adwa",
      amharic: "ቅዱስ ጊዮርጊስ - የአድዋ ጦርነት",
    },
    description: {
      english: "Saint Giorgis (George)-The Battle of Adwa",
      amharic: "ቅዱስ ጊዮርጊስ - የአድዋ ጦርነት",
    },
    recuringDate: 23,
    major: true,
    negs: 6,
  },

  takla_haymanot: {
    // Information about the saint is defined ONCE
    name: {
      english: "Abuna Takla Haymanot",
      amharic: "አቡነ ተክለ ሐይማኖት",
    },
    recuringDate: 24,
    defaultDescription: {
        english: "Commemoration of Abuna Takla Haymanot.",
        amharic: "የአቡነ ተክለ ሐይማኖት መታሰቢያ።",
    },
    events: [
      {
        key: "takla_haymanot_diparted_from_this_world",
        description: {
          english: "Abuna Takla Haymanot departed from this world",
          amharic: "አቡነ ተክለ ሃይማኖት ከዚህ ዓለም በሞት ተለዩ",
        },
        major: true,
        negs: 12, // The major annual feast for his passing is in the 12th month
      },
      {
        key: "takla_haymanot_legs_broke",
        description: {
          english: "Having stood for 7 years, one of his legs broke",
          amharic: "ለሰባት ዓመታት ከቆሙ በኋላ አንደኛው እግራቸው ተሰበረ።",
        },
        major: true, // `major` is now tied to the specific event
        negs: 5,   // The major feast for this event is in the 5th month
      },
      {
        key: "takla_haymanot_birth_date",
        description: {
          english: "Abuna Takla Haymanot's birth date",
          amharic: "የአቡነ ተክለ ሃይማኖት የልደት ቀን",
        },
        major: false,
        negs: 4,   // The major feast for his birth is in the 4th month
      },
    ],
  },
  saint_kirstos_semera: {
    key: "saint_kirstos_semera",
    name: {
      english: "Saint Kirstos Semera",
      amharic: "ቅድስት ክርስረቶስ ሰምራ ",
    },
    description: {
      english: "Saint Kirstos Semera",
      amharic: "ቅድስት ክርስረቶስ ሰምራ ",
    },
    recuringDate: 24,
    major: true,
    negs: 12,
  },
  heavenly_priests: {
    key: "24_heavenly_priests",
    name: {
      english: "24 Heavenly Priests",
      amharic: "24ቱ ካህናተ ሰማይ",
    },
    description: {
      english: "24 Heavenly Priests",
      amharic: "24ቱ ካህናተ ሰማይ",
    },
    recuringDate: 24,
    major: true,
    negs: 3,
  },

  saint_marqorewos: {
    key: "saint_marqorewos",
    name: {
      english: "Saint Marqorewos (Merkorios)/ Abune Habibe",
      amharic: "ቅዱስ መርቆሬዎስ ፣አቡነ ሀቢብ (አባ ቡላ)",
    },
    description: {
      english: "Saint Joseph",
      amharic: "አረጋዊው ዮሴፍ ",
    },
    recuringDate: 27,
    major: true,
    negs: 3,
  },
  abune_habibe: {
    key: "abune_habibe",
    name: {
      english: "Abune Habibe",
      amharic: "አቡነ ሀቢብ (አባ ቡላ)",
    },
    description: {
      english: "Abune_Habibe",
      amharic: "አረጋዊው ዮሴፍ ",
    },
    recuringDate: 25,
    major: true,
    negs: 2,
  },
  saint_jude_son_of_alpheus: {
    key: "saint_jude_son_of_alpheus",
    name: {
      english: "Saint Jude, son of Alpheus, the Apostle",
      amharic: "ቅዱስ ይሁዳ ዘአልፍዮስ ሐዋርያው",
    },
    description: {
      english: "Saint Jude, son of Alpheus, the Apostle his martyrdom",
      amharic: "ቅዱስ ይሁዳ ዘአልፍዮስ ሐዋርያውሰማዕትነቱ",
    },
    recuringDate: 25,
    major: false,
    negs: 10,
  },

  saint_joseph: {
    key: "saint_joseph",
    name: {
      english: "Saint Joseph ",
      amharic: "አረጋዊው ዮሴፍ",
    },
    description: {
      english: "Saint Joseph",
      amharic: "አረጋዊው ዮሴፍ ",
    },
    recuringDate: 26,
    major: true,
    negs: 11,
  },
  abba_salama_frumentius: {
    key: "abba_salama_frumentius",
    name: {
      english: " Abba_Salama_Frumentius, the Enlightener of Ethiopia",
      amharic: "አባ ሰላማ ከሳቴ ብርሀን",
    },
    description: {
      english: "Abba_Salama_Frumentius, the Enlightener of Ethiopia",
      amharic: "አባ ሰላማ ከሳቴ ብርሀን ",
    },
    recuringDate: 26,
    major: true,
    negs: 11,
  },
  thomas_the_apostle: {
    key: "thomas_the_apostle",
    name: {
      english: "Thomas the apostle",
      amharic: " ቶማስ ዘህንደኬ",
    },
    description: {
      english: "Thomas the apostle",
      amharic: "ቶማስ ዘህንደኬ ",
    },
    recuringDate: 26,
    major: true,
    negs: 9,
  },
  abune_habte_mariam: {
    key: "abune_habte_mariam",
    name: {
      english: "Abune Habte Mariam",
      amharic: "አቡነ ሀብተ ማርያም",
    },
    description: {
      english: "Abune Habte Mariam",
      amharic: "አቡነ ሀብተ ማርያም",
    },
    recuringDate: 26,
    major: true,
    negs: 3,
  },
  aba_eyesus_moea: {
    key: "aba_eyesus_moea",
    name: {
      english: "Aba Eyesus Moe`a",
      amharic: "አባ ኢየሱስ ሞአ",
    },
    description: {
      english: "Saint Eyesus Moe`a",
      amharic: "አባ ኢየሱስ ሞአ ",
    },
    recuringDate: 26,
    major: true,
    negs: null,
  },

  medhane_alem: {
    key: "medhane_alem",
    name: {
      english: "Medhane Alem -The Saviour of the world/ Abune Mebea Zion",
      amharic: "መድሐኔአለም ",
    },
    description: {
      english: "Medhane Alem -The Saviour of the world",
      amharic: "የዓለም መድኃኒት",
    },
    recuringDate: 27,
    major: true,
    negs: 7,
  },
  abune_mebea_zion: {
    key: "abune_mebea_zion",
    name: {
      english: " Abune Mebea Zion",
      amharic: "አቡነ መባዓ ፅዮን",
    },
    description: {
      english: "Abune Mebea Zion",
      amharic: "አቡነ መባዓ ፅዮን",
    },
    recuringDate: 27,
    negs: 2,
  },

  ammanuel: {
    key: "ammanuel",
    name: {
      english: "ammanuel",
      amharic: "አማኑኤል",
    },
    description: {
      english: "ammanuel",
      amharic: "አማኑኤል",
    },
    recuringDate: 28,
    major: true,
    negs: [2, 5, 9],
  },

  baale_wold: {
    key: "baale_wold",
    name: {
      english: "Ba`ale Wold",
      amharic: "በዓለወልድ",
    },
    description: {
      english: "Ba`ale Wold (Feast of God The Son)",
      amharic: "ዮሐንስ መጥምቅ",
    },
    recuringDate: 29,
    major: true,
    negs: [4, 5],
  },
  saint_lalibela: {
    key: "saint_lalibela",
    name: {
      english: "Saint Lalibela",
      amharic: "ቅዱስ ላሊበላ",
    },
    description: {
      english: "Saint Lalibela",
      amharic: "ቅዱስ ላሊበላ",
    },
    recuringDate: 29,
    major: true,
    negs: 4,
  },
  saint_arsema_Beheading: {
    key: "saint_arsema_29",
    name: {
      english: "Saint Arsema",
      amharic: "አርሴማ",
    },
    description: {
      english: "Commemoration of The Martyrdom (Beheading) of Kidist Arsema",
      amharic: "የቅድስት አርሴማ ሰማዕትነት (ራስ መቁረጥ)",
    },
    recuringDate: 29,
    major: false,
    negs: 1,
  },

  john_baptist: {
    key: "john_baptist",
    name: {
      english: "John the baptist",
      amharic: "ዮሐንስ መጥምቅ ",
    },
    description: {
      english: "John the baptist",
      amharic: "ዮሐንስ መጥምቅ",
    },
    recuringDate: 30,
    major: true,
    negs: 10,
  },
  saint_markos: {
    key: "saint_markos",
    name: {
      english: "Saint_Markos",
      amharic: "ቅዱስ ማርቆስ",
    },
    description: {
      english: "Saint Markos /St. Mark the Evangelist/",
      amharic: "የቅዱስ ማርቆስ የልደት",
    },
    recuringDate: 30,
    major: true,
    negs: [2, 8],
  },
};
